### PR TITLE
Allow a customer to perform further action for a Stripe payment if needed (#4181)

### LINKED ISSUE #4181: Allow a customer to perform further action for a Stripe payment if needed
## Description

- Relates to #4170 

**- As a:** customer
**- On page:** Payments Requiring Action
**- I want to be able to do:** Handle authentication needed for an existing Stripe payment attempt if needed

See [`next_action.redirect_to_url`](https://stripe.com/docs/api/payment_intents/object#payment_intent_object-next_action-redirect_to_url) in the Stripe `PaymentIntent` object.

## Acceptance Criteria

The customer should be able to handle authentication needed for a subscription order if payment failed because it needs SCA authentication.

This should cover payments from Subscriptions and also Payments triggered by the manager when collecting payments in the backoffice. See discussion in #4173 for details.

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Closes #4181

If further action is required to complete a payment, we want the customer to be able to take that action from the transactions tab on the Account page. 

#### What should we test?

- Create a payment that requires additional authorization using card x-3184, either via admin payment or a subscription.
- Verify that the payment appears as pending on the Transactions tab
- Click the link and authorize the payment
- The payment should now show as completed 

#### Release notes

Added a link to confirm pending transactions on the Account page

Changelog Category: User facing changes

#### Dependencies

https://github.com/openfoodfoundation/openfoodnetwork/pull/6713

### CODE REVIEW

File: app/serializers/api/order_serializer.rb
Comment (Matt-Yorkley) [inline app/serializers/api/order_serializer.rb:19]: There's a failing test in the order serializer spec, I'm assuming it's due to this line.

**Comment by Matt-Yorkley** [comment]
@andrewpbrett I'm wondering if we might need a new "state" in the payments model. It feels a bit like there are some square-peg-round-hole type issues in trying to fit the logic we need into the existing states. Like when a payment has been set up but is awaiting authorization, it feels like it's not quite "pending" and it's not quite "complete", but something in between which isn't fully encapsulated by either state. What do you think?

**Comment by andrewpbrett** [comment]
@Matt-Yorkley Yeah that's not a bad point. A couple times when working on this I had the thought that it should be a separate state. I didn't go for it, mainly because I didn't want to introduce any more complexity to the payment state machine and because it didn't feel like it was absolutely necessary. 

When I sketched out what it would look like, it was essentially a state with a different name that, for the purposes of the state machine, acted the same as the "pending" state. 

The two differences were 1) we could name it differently, so like "authorization required" or something, and 2) we had a url for the customer to authorize it. 

Difference 1) didn't feel like it was sufficient (in terms of language, "pending" captures it decently well already), and difference 2) is reflected in the `cvv_response_message` field. The payment requires authorization if and only if isn't `nil`, since once the customer authorizes it, we clear the field. 

I might be missing something with how a "pending" payment behaves or is treated, where we wouldn't want an "auth required" payment to behave/be treated in the same way. But I thought they were equivalent enough where having less code outweighed the benefit of another state. 

I'll take a look at that failing spec, too 👍

**Comment by Matt-Yorkley** [comment]
I was wondering about the behavior on the "capture" buttons for pending payments. If it's pending it can be manually captured, right? But maybe that doesn't make sense in this case. Maybe there's just a small UX tweak that could be made without adding a new state :shrug:

Anyway, avoiding adding a new state seems preferable if it's not really needed :ok_hand:

**Comment by andrewpbrett** [comment]
If there's a `cvv_response_message` present, we show a different button and the capture button isn't available:

Screen Shot 2021-02-05 at 10 17 45 AM

That's from a different PR, I think that piece might already be deployed actually.

File: app/serializers/api/order_serializer.rb
Comment (andrewpbrett) [reply app/serializers/api/order_serializer.rb]: 9aecefb05d24250e0168b99cf6ea53e54ebe393f - I looked around for other places that changing this might impact but I didn't see any that we're explicitly expecting pending payments to **not** be returned.

**Review by Matt-Yorkley — APPROVED** [review]

File: app/serializers/api/order_serializer.rb
Comment (sauloperez) [inline app/serializers/api/order_serializer.rb:19]: just a matter of styling, I believe `.where(state: %w(completed pending))` would also work and doesn't require you to get down to SQL. It's a matter of preference, so feel free to dismiss.

File: app/views/spree/users/_fat.html.haml
Comment (sauloperez) [inline app/views/spree/users/_fat.html.haml:20]: There's something a bit confusing from the resulting UI (notice I faked the payment a bit. I just provided a `cvv_response_message` on an existing payment). On that page, we highlight the balance in red but it also indicates a link.
Screenshot from 2021-02-10 17-07-55

Wouldn't it be more understandable for the user to highlight the required action a bit more somehow? What are your thoughts on this @Erioldoesdesign ?  I wonder if users will notice they need to click there... we check that in production though and improve it in another PR.

**Review by sauloperez — COMMENTED** [review]
Just a question, how do users know they need to authorize the payment from that tab? Is it a link from one of the emails we send? I couldn't find it.

File: app/views/spree/users/_fat.html.haml
Comment (sauloperez) [reply app/views/spree/users/_fat.html.haml:20]: Talked offline with  @andrewpbrett and this is just a fallback from the email notification users get. If they lose the email they can come here and still authorize the payment.

**Review by sauloperez — APPROVED** [review]

File: app/views/spree/users/_fat.html.haml
Comment (jibees) [reply app/views/spree/users/_fat.html.haml:20]: I agree with you @sauloperez: this is not obvious that this link is clickable. And its more a CTA than a simple link (like `Order R25...`). Maybe we should have _small button components_ to meet this need.

Comment (sauloperez) [reply app/views/spree/users/_fat.html.haml:20]: yes, I feel we need to explore components that are neither CTAs nor plain old links :ok_hand: I hope this fits into feature discussions about a style guide.

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

I've looked at this scenario before staging, and realized we might have an issue with the payment state of orders. Basically, payments placed in the backoffice need to be manually captured, after the authentication step. I think this might be a bug, right?
Opened issue https://github.com/openfoodfoundation/openfoodnetwork/issues/6867.

I staged this PR, and checked your "what to test" section - thanks for providing that. So, going stepwise for **Backoffice orders**:

i) As admin, created a new order - via `/admin/orders/new` - and created a payment in that order, which requires additional authorization (using card `x-3184`) -> **OK**
ii) As admin, verified that the payment appears as pending on the Transactions tab -> **OK**
iii) As a customer, clicked the link and authorize the payment -> **I could not find that link under** `/account#/transactions#`. I've placed several orders on the BO, but in no case the order appeared on `/transactions`, with a link to redirect to the authentication page... This is the expected behavior here, right? Adding the feedback label.

image


Continuing:
iv) However, the link arrives per email, which can then be used by the customer to provide authorization -> doing so, leads us to an order confirmation page, which is marked as "not paid" -> I think this is issue #6867? 

image

v) even after providing authorization, the order does not appear under the transactions tab for the customer, and is still pending for the admin

vi) it is only after the admin manually visits `/admin/orders/R184115465/edit` the order marked as "complete" and appears on the transactions tab of the customer account, marked as "paid". At this point, the order confirmation emails arrive.

**Subscriptions**

After placing a subscription, the payment is automatically marked as failed, but the order appears as "complete". Since the order is complete, it also appears under the **transactions** tab, for the customer:

image

 
Should this still be worked on? I've spotted no regression, but I'm not sure it should be merged as is...

**Comment by andrewpbrett** [comment]
@filipefurtad0 thanks for taking a look at this. I should have been more clear - this relies on the changes in #6713 to work properly. I believe #6713 will also address what you (correctly!) spotted and documented in #6867 :)

**Comment by filipefurtad0** [comment]
Hey @andrewpbrett ,

So, this issue is blocked by #6713, is that right? In that case, they should be have been tested in the inverse order: first, #6713 and - only after merging - the present issue #6808? Is this correct?

**Comment by andrewpbrett** [comment]
@filipefurtad0 Yes that's correct. Sorry about that! :-/

File: app/services/process_payment_intent.rb
Comment (sauloperez) [inline app/services/process_payment_intent.rb:10]: :clap: :clap:

File: app/controllers/spree/orders_controller.rb
Comment (sauloperez) [inline app/controllers/spree/orders_controller.rb:31]: I'm missing controller tests for this new payment-intent behavior.

**Comment by filipefurtad0** [comment]
Following up on your [comment](https://github.com/openfoodfoundation/openfoodnetwork/issues/6867#issuecomment-778475014):
Thanks for addressing this @andrewpbrett.

I think we need an automated test to cover issue #6867 (backoffice), which is basically a repetition of #6755 (shopfront).  

The test case we want to cover is basically, something like:
- authorizing a stripe SCA payments, from valid cards
i) triggers order confirmation emails
ii) changes the order state from "Payment" to complete
iii) automatically capture payment

I realize this needs some thinking. Should we pair up on this?

File: app/serializers/api/order_serializer.rb
Comment (sauloperez) [reply app/serializers/api/order_serializer.rb:19]: Now that I checked again, we need payments in `pending` state so we can show them as requiring auth in /account, right?

File: app/services/process_payment_intent.rb
Comment (sauloperez) [inline app/services/process_payment_intent.rb:22]: this new extra behavior should be reflected in this class' unit tests IMO

**Review by sauloperez — CHANGES_REQUESTED** [review]

**Comment by sauloperez** [comment]
sorry for my negative re-review but I really think we need to extend our tests. I can help you on that helps you unblock this.

File: app/services/process_payment_intent.rb
Comment (andrewpbrett) [reply app/services/process_payment_intent.rb:22]: As a start, I added some for the class here: 89123428c605006c45ae4f030dd588a17ed2a11c - could probably be improved upon though.

File: app/controllers/spree/orders_controller.rb
Comment (andrewpbrett) [reply app/controllers/spree/orders_controller.rb:31]: 05e762ff2151c0fd5406f000a8ba3795f28fe7d3

**Comment by andrewpbrett** [comment]
No need to apologize @sauloperez, your feedback is 100% correct. I think it should be addressed now but let me know if you spot any places it could be improved.

File: app/serializers/api/order_serializer.rb
Comment (andrewpbrett) [reply app/serializers/api/order_serializer.rb:19]: Yep, that's correct.

**Review by sauloperez — APPROVED** [review]
This is perfect now! :clap:

**Comment by filipefurtad0** [comment]
Cool @andrewpbrett ,

We're getting closer to solving this one.

**regarding #6867** After creating an order and a payment that requires additional authorization using card x-3184 in the BO, we get to this state:

image

The order is in a "payment" state, waiting for the customer to provide authorization, right? So far so good.

We need to consider, that the admin will probably not sit in this screen, while the customer takes his time to go through his emails, and finally getting his auth-step complete. So if the admin leaves the "Payments" tab, and looks for this order on `/orders` it will look like this:

image

So, I guess it's quite tempting to click that check-sign - and this triggers this gateway error:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/602d18f5a32bd8001865dc83?event_id=602d462d006996ddac860000&i=sk&m=nw

This renders the payment status "Failed", and I believe a new payment is necessary.

The good news is, that if the admin does not try to capture that payment, and the customer provides authentication through the email, then, the order reaches a "complete" state :tada: So this is an improvement, it was not happening before - before we needed action from the admin.

**regarding #4181** and what this PR aims to achieve - same starting point: After creating an order and a payment that requires additional authorization using card x-3184 in the BO I still can't observe the link, on the user's transaction tab, as [displayed on this comment](https://github.com/openfoodfoundation/openfoodnetwork/pull/6808#discussion_r573862645). Please see the video, in which the order is created in the BO, and the customer's pages is refreshed after that - no sign of that new order:

Peek 2021-02-17 16-44

This PR improves #6867 and brings new tests so - I think we should:
- merge this PR
- keep #4181 open
- eventually downgrade #6867

What do you think?

I think we should this issue as in mind, as it looks related:
Unsuccessful SCA payments repeatedly introduce transaction fees (backoffice) 
https://github.com/openfoodfoundation/openfoodnetwork/issues/6683

**Comment by filipefurtad0** [comment]
Please notice @andrewpbrett , that we are good on subs:

- after running `bundle exec rake ofn:subs:test:repeat_placement_job`

image

- after running `bundle exec rake ofn:subs:test:force_confirmation_job`

In the case of subs, the pending payment appears, as well as the link to provide authorization :tada: 

image

- after the customer proceeds with authentication, all is good:

image

The difference in the payment **state**, while waiting for authentication:
- in the case of backoffice orders -> placed order is in the state `payment` -> **not appearing** under `/transactions` on the customers side
- in the case of orders placed by subscriptions -> placed order is in the state `complete` -> **appears** under `/transactions` on the customers side

Hope this helps, but again - let's merge this!

**Comment by andrewpbrett** [comment]
Noted and merged, thanks @filipefurtad0! I'll take a look at what could be happening there.

File: app/views/spree/users/_fat.html.haml
Comment (andrewpbrett) [reply app/views/spree/users/_fat.html.haml:20]: I made them buttons for now: #6919

**Pull request merged**
